src/github.com/matrix-org/dendrite/roomserver/storage/storage.go

```diff
@@ -381,3 +379,128 @@ func (d *Database) StateEntriesForTuples(
 ) ([]types.StateEntryList, error) {
 	return d.statements.bulkSelectFilteredStateBlockEntries(stateBlockNIDs, stateKeyTuples)
 }
+
+type membershipUpdater struct {
+	transaction
+	d          *Database
+	roomNID    types.RoomNID
+	targetNID  types.EventStateKeyNID
+	membership membershipState
+}
+
+func (d *Database) membershipUpdaterTxn(
+	txn *sql.Tx, roomNID types.RoomNID, targetNID types.EventStateKeyNID,
+) (types.MembershipUpdater, error) {
+
+	if err := d.statements.insertMembership(txn, roomNID, targetNID); err != nil {
+		return nil, err
+	}
+
+	membership, err := d.statements.selectMembershipForUpdate(txn, roomNID, targetNID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &membershipUpdater{
+		transaction{txn}, d, roomNID, targetNID, membership,
+	}, nil
+}
+
+// IsInvite implements types.MembershipUpdater
+func (u *membershipUpdater) IsInvite() bool {
+	return u.membership == membershipStateInvite
+}
+
+// IsJoin implements types.MembershipUpdater
+func (u *membershipUpdater) IsJoin() bool {
+	return u.membership == membershipStateJoin
+}
+
+// IsLeave implements types.MembershipUpdater
+func (u *membershipUpdater) IsLeave() bool {
+	return u.membership == membershipStateLeaveOrBan
+}
+
+// SetToInvite implements types.MembershipUpdater
+func (u *membershipUpdater) SetToInvite(event gomatrixserverlib.Event) (bool, error) {
+	// TODO: assign the state key inside the transaction.
+	senderNID, err := u.d.assignStateKeyNID(event.Sender())
+	if err != nil {
+		return false, err
+	}
+	eventID := event.EventID()
```

Comment:
Why aren't you inlining this? You never use this variable anywhere else. In this form, it just adds to my cognitive load as I have to parse each function and consider where you might update it / use it elsewhere. Inlining it removes all those considerations.